Add `TokenMemoryCache`

gretchenfrage opened the pull request:
Implements `TokenMemoryCache`, and makes it the default `TokenStore` for clients.

Ralith commented on quinn-proto/Cargo.toml:
```diff
@@ -56,6 +56,7 @@ web-time = { workspace = true }
 
 [dev-dependencies]
 assert_matches = { workspace = true }
+rand_pcg = "0.3"
```
Please update to latest.

Ralith commented on quinn-proto/src/token_memory_cache.rs:
```diff
@@ -0,0 +1,539 @@
+/// `TokenStore` implementation that stores up to `N` tokens per server name for up to a
+/// limited number of server names, in-memory
+#[derive(Debug)]
+pub struct TokenMemoryCache<const N: usize>(Mutex<State<N>>);
```
Does `N` really need to be a compile-time constant? Dynamic values are simpler and mean less work for the compiler downstream.

gretchenfrage replied:
Yeahhhhhhhhhhh I guess not. My argument of avoiding a heap allocation seems kinda flimsy since there's already going to be 1 mandatory heap allocation for the server name and then another for each token.

Ralith replied:
Heap allocations aren't something to be avoided for their own sake, in any case.

Ralith commented on quinn-proto/src/token_memory_cache.rs:
```diff
@@ -0,0 +1,539 @@
+
+/// In-place vector queue of up to `N` `Bytes`
+#[derive(Debug)]
+struct Queue<const N: usize> {
```
Would it be simpler for this to be a `VecDeque`?

gretchenfrage commented on quinn-proto/src/config/mod.rs:
```diff
@@ -585,8 +585,8 @@ impl ClientConfig {
     ///
-    /// Defaults to [`NoneTokenStore`], which disables the use of tokens from NEW_TOKEN frames as a
-    /// client.
+    /// Default to [`TokenMemoryCache`] with 2 tokens per server, which is suitable for most
```
`Default` -> `Defaults`.

gretchenfrage commented on quinn-proto/src/token_memory_cache.rs:
```diff
@@ -0,0 +1,539 @@
+
+/// Slab-based linked LRU cache of `CacheEntry`
+#[derive(Debug, Default)]
+struct LinkedCache<const N: usize> {
```
I like it :D I'm worried about this though https://github.com/Ralith/lru-slab/issues/1

Ralith replied:
Fixed in 0.1.2. Not sure anyone should be storing billions of tokens in memory regardless.

gretchenfrage replied:
Increased dependency version in this PR.

gretchenfrage commented:
Switched from bespoke LRU cache implementation to Ralith's `lru_slab` crate. Saves a lot of code for sure, although I don't wanna merge until we decide what we wanna do about https://github.com/Ralith/lru-slab/issues/1

Switched from bespoke in-place queue implementation to just using `VecDeque`.

Together, these changes cut the size of this PR roughly in half. Nice!

Ralith commented on quinn-proto/src/token_memory_cache.rs:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    fn take(&mut self, server_name: &str) -> Option<Bytes> {
+        if let hash_map::Entry::Occupied(hmap_entry) = self.lookup.entry(server_name.into()) {
```
style nit: use let-else to reduce rightward drift

Ralith commented on quinn-proto/src/token_memory_cache.rs:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    fn take(&mut self, server_name: &str) -> Option<Bytes> {
+        if let hash_map::Entry::Occupied(hmap_entry) = self.lookup.entry(server_name.into()) {
```
Here you're allocating a fresh `Arc<str>` and using a less common API to avoid performing multiple lookups into the hash table in the rare case of token exhaustion. Is that worthwhile?

gretchenfrage replied:
Probably not--unless the user set tokens_per_server to 1. I'll switch it.

gretchenfrage commented:
@djc do you want to review this as well? It's fine either way.

djc commented:
Yes, will try to take a look soon.

djc commented on quinn-proto/src/config/mod.rs:
```diff
@@ -585,8 +585,8 @@ impl ClientConfig {
     ///
-    /// Defaults to [`NoneTokenStore`], which disables the use of tokens from NEW_TOKEN frames as a
-    /// client.
+    /// Defaults to a default [`TokenMemoryCache`], which is suitable for most internet
```
Nit: "Defaults to a default" seems a little ugly? I'd probably omit "a default", since I think that is the reasonable assumption anyway.

djc reviewed (approved):
Nice!
